generate_bear_requirements.py: Add language file

sladyn98 opened the pull request:
This generates a yaml file mapping the languages used by every bear.

Closes https://github.com/coala/coala-bears/issues/2479

<!--
Thanks for your contribution!

Please take a quick look at those things down there. They're quite important.
Really! We wrote them for you. Yes you! With utmost care. Read them.
-->

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `corobo mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

shikharvaish28 commented on .ci/generate_bear_requirements.py:
```diff
@@ -163,6 +164,20 @@ def get_cabal_requirements(requirements):
 
+def get_languages(bears):
+    language_dict={}
+<<<<<<< HEAD
```
You need to solve this merge conflict

sladyn98 replied:
Sould i squash all commits

shikharvaish28 replied:
First of all, try to solve the conflict and then squash it. (In case you're not able to find where to solve the issue, try to squash them all, it will raise a merge conflict error which you can solve from your editor)

li-boxuan commented on .ci/generate_bear_requirements.py:
```diff
@@ -163,6 +164,13 @@ def get_cabal_requirements(requirements):
+def get_languages(bears):
+    language_dict={}
+    for bear in bears:
+        language_dict[str(bear.name)]=bear.LANGUAGES
```
redundant `str()` ?

li-boxuan replied:
I think it returns a string literal. Correct me if I am wrong.

sladyn98 replied:
When i coded the solution and tested  it gave me an error so i put it there.

li-boxuan commented on .ci/generate_bear_requirements.py:
```diff
@@ -258,6 +266,11 @@ def sort_requirements(req_dict):
+    language_requirements = get_languages(get_all_bears(bear_dirs))
+    file_path = os.path.join(PROJECT_DIR,BEAR_LANGUAGE_REQUIREMENTS_YAML)
+    with open(file_path,'w') as outfile:
+        yaml.dump(language_requirements,outfile)
```
Better sort them before dumping

sladyn98 commented on .ci/generate_bear_requirements.py:
```diff
@@ -258,6 +266,11 @@ def sort_requirements(req_dict):
     else:
         output = open(args.output, 'w')
 
+    language_requirements = get_languages(get_all_bears(bear_dirs))
```
I need the language function to create a dictionary to dump in yaml.

li-boxuan replied:
I mean you can reuse the result of this `get_all_bears(bear_dirs)`. See comment above.

li-boxuan replied:
This is not addressed.

sladyn98 replied:
@li-boxuan how do i squash them i cant access them from my terminal

sladyn98 replied:
I got it thanks i learnt a lot from this. It was because i kept pushing

ayan-b replied:
Glad it helped. :blush:

li-boxuan replied:
Again, this is not addressed. If you disagree with me, you can comment here. Otherwise please make changes.

Dhiraj240 commented:
@sladyn98 Its a basic requirement , you can adjust things like ``` Generate language file ```  as ```Add file or Add language```

li-boxuan commented on bear-requirements.yaml:
```diff
@@ -172,10 +172,6 @@ pip_requirements:
     version: ~=3.4.0
   lxml:
     version: ==3.6.0
-  memento_client:
```
But it looks like those libraries are being used. Could you post the command how you get these removals?

Even if the removal is necessary, IMO you should create an issue instead of solving it in this pr. At least you should split it into another commit, since this is unrelated to your current commit.

li-boxuan replied:
This is not addressed.

jayvdb commented on bear-language-requirements.yaml:
```diff
@@ -0,0 +1,355 @@
+ApertiumLintBear: !!set
```
we do not need these to be sets.

simple lists would be ok.

the set is coming from `bear.LANGUAGES`.  You should convert that to a list and sort it.

sladyn98 replied:
I made the requested changes @jayvdb

li-boxuan replied:
This is not addressed.

Naveenaidu commented on .ci/generate_bear_requirements.py:
```diff
@@ -87,6 +88,7 @@ def get_all_bears(bear_dirs):
     return list(itertools.chain(local_bears, global_bears))
 
 
+
```
This extra line is unnecessary

sladyn98 replied:
made the changes

Naveenaidu commented on .ci/generate_bear_requirements.py:
```diff
@@ -163,6 +165,13 @@ def get_cabal_requirements(requirements):
+        language_dict[str(bear.name)]=list(bear.LANGUAGES).sort()
+    return language_dict
+
+
```
This is unnecessary too.

sladyn98 commented:
@li-boxuan i made the requested changes by adding those files back.

li-boxuan commented on bear-language-requirements.yaml:
Please also sort bears according to their names.

ishanSrt commented:
Now I notice, many of the things that I pointed out have already been addressed by @li-boxuan but you didn't make the changes. Please address all your review before asking for more. 👍

ishanSrt commented on .ci/generate_bear_requirements.py:
```diff
@@ -258,6 +268,12 @@ def sort_requirements(req_dict):
         output = open(args.output, 'w')
 
+    language_requirements = get_languages(get_all_bears(bear_dirs))
+    file_path = os.path.join(PROJECT_DIR,BEAR_LANGUAGE_REQUIREMENTS_YAML)
```
space after comma

ishanSrt replied:
is Gitmate bot taking a nap? 😅

sladyn98 replied:
Yes i made these changes.

li-boxuan commented on tests/swift/TailorBearTest.py:
```diff
@@ -24,22 +24,22 @@
 }
 """
 
-TailorBearWithoutConfig = verify_local_bear(TailorBear,
```
Seems these changes are from 7bc134e59873137fb1b9f0feacea3bce77d2ed80. Please use git properly.

sladyn98 replied:
@li-bouxuan can you kindly explain how do I get rid of this issue.

li-boxuan replied:
You encountered this issue because you didn't rebase your commit properly. To solve this, revert your changes on `TailorBearTest.py` and `TSLintBearTest.py`.

ishanSrt commented on .ci/generate_bear_requirements.py:
```diff
@@ -26,6 +26,8 @@
 from coalib.bears.BEAR_KIND import BEAR_KIND
 from coalib.collecting.Collectors import collect_bears
 
+from pyprint.ConsolePrinter import ConsolePrinter
```
unused import

sladyn98 replied:
I removed this import

ishanSrt commented on bear-language-requirements.yaml:
```diff
@@ -0,0 +1,411 @@
+ApertiumLintBear:
```
Please run YAMLLintBear on this file, since YAML is guided by indentation, I see inconsistent styles being used for the bears with no dependencies i.e. with `[]` in comparison to the rest of the file, which can be avoided easily

ishanSrt replied:
The changes have been made by @sladyn98. Marking as resolved.

sladyn98 commented on .travis.yml:
```diff
@@ -4,7 +4,6 @@ language: python
 python:
   - 3.4
   - 3.5
-  - 3.6
```
This change was not done by me. Should i change it back

anctartica replied:
yes please

sladyn98 commented on .ci/generate_bear_requirements.py:
```diff
@@ -162,6 +166,12 @@ def get_pip_requirements(requirements):
+def get_languages(bears):
+    language_dict = {}
+    for bear in bears:
+        language_dict[str(bear.name)]=list(sorted(bear.LANGUAGES))
```
Done

sladyn98 commented on .ci/generate_bear_requirements.py:
```diff
@@ -258,6 +268,12 @@ def sort_requirements(req_dict):
     else:
         output = open(args.output, 'w')
 
+    language_requirements = get_languages(get_all_bears(bear_dirs))
```
Yes I solved this

anctartica commented on .ci/generate_bear_requirements.py:
```diff
@@ -258,6 +268,12 @@ def sort_requirements(req_dict):
+    language_requirements = get_languages(get_all_bears(bear_dirs))
+    file_path = os.path.join(PROJECT_DIR,BEAR_LANGUAGE_REQUIREMENTS_YAML)
+    with open(file_path, 'w') as outfile:
+        yaml.indent(mapping = 2, sequence = 4, offset = 2)
```
is this required? the previous code doesn't need to use this.

sladyn98 commented:
@anctartica yes it is because i need proper indentation.

sladyn98 commented:
@li-boxuan  @ishanSrt  I made all the changes that you guys requested. Except i cant figure out about the test lint bear, because those changes have already been merged to master, they cant be removed.
Apart from that if any changes are required kindly suggest.

jayvdb commented:
Needs to be rebased properly, and commit message *must* follow https://coala.io/commit .  These are basics that *must* be done, otherwise CI fails, and it is not worth reviewer time to investigate problems with your PR because it is very probable that the problem is that you are using old code , not rebased.

sladyn98 commented:
@jayvdb I did make the requested chages. I rebased, I changed the commit message, I made the changes requested by @li-boxuan  and @ishanSrt thank you for teaching me these valuable things. The gitmate bot is giving me an error on some files but i guess that is a bug.

li-boxuan commented on .coafile:
```diff
@@ -82,4 +82,4 @@ shellcheck_ignore = SC2046, SC2154, SC2155
 files = **.jj2
 # Bug in handling of ".. list | sort"
-ignore = .moban.dt/Gemfile.jj2, .moban.dt/bear-requirements.txt.jj2, .moban.dt/package.json.jj2
+ignore = .moban.dt/Gemfile.jj2, .moban.dt/bear-requirements.txt.jj2, .moban.dt/package.json.jj2
```
Please remove unrelated change.

sladyn98 commented:
@li-boxuan dont you think that is a bug because in the original  coala repo the file is the same.

li-boxuan commented:
The issue requires you to generate a mapping of languages to bears, but you are generating a mapping of bears to languages. You need to either explain on the issue page, or you need to follow the requirement.

sladyn98 commented:
@li-bouxan i spoke to John about this and he said it's fine.

sladyn98 commented:
@li-boxuan I addressed the chaneges the boxes are no longer there

sladyn98 commented:
@li-bouxan I made the changes.

jayvdb commented:
Rebase your patch before submitting it for review.

jayvdb commented:
Please look at the Travis CI failures.  They show an obvious problem with your patch.

jayvdb commented on bear-language-requirements.yaml:
```diff
@@ -0,0 +1,411 @@
+  - c
+  - cpp
+  - cs
+  - default
```
Interesting ... `default` is not a language.  Maybe this is a bug in the bear, or this PR needs to create a special case to skip that one.

jayvdb commented on bear-language-requirements.yaml:
again, 'default'

jayvdb commented on bear-language-requirements.yaml:
'All' isnt a language.  It should be omitted from the yaml file.  This should be a special case.

maybe write these as

```yaml
KeywordBear: true
```

to indicate the bear exists, but has no defined languages.
`KeywordBear: false` is also a possible way of describing these bears.

sladyn98 commented:
> Please look at the Travis CI failures. They show an obvious problem with your patch.
git diff is exiting with code 1

DanPi32 commented on bear-language-requirements.yaml:
```diff
@@ -0,0 +1,411 @@
+  - C++
+  - CUDA
+  - Objective-C
+  - Objective-C++
```
```suggestion
  - Objective-CPP
```

DanPi32 commented on bear-language-requirements.yaml:
```diff
@@ -0,0 +1,411 @@
+  - C++
+  - CUDA
+  - Objective-C
+  - Objective-C++
```
```suggestion
  - Objective-CPP
```

DanPi32 commented on bear-language-requirements.yaml:
```diff
@@ -0,0 +1,411 @@
+  - C++
+  - CUDA
+  - Objective-C
+  - Objective-C++
```
```suggestion
  - Objective-CPP
```

DanPi32 commented on bear-language-requirements.yaml:
```diff
@@ -0,0 +1,411 @@
+  - C++
+  - CUDA
+  - Objective-C
+  - Objective-C++
```
```suggestion
  - Objective-CPP
```

DanPi32 commented on bear-language-requirements.yaml:
```diff
@@ -0,0 +1,411 @@
+  - C++
+  - CUDA
+  - Objective-C
+  - Objective-C++
```
```suggestion
  - Objective-CPP
```

sladyn98 commented:
@jayvdb  Changed the commit message.

jayvdb commented:
no, you edited the PR page on github.

read https://coala.io/commit and https://coala.io/git , or stackoverflow for amend git commit etc.

sladyn98 commented:
@jayvdb Yes now i used git commit --amend and made the changes.Thanks and kindly review

jayvdb commented on bear-language-requirements.yaml:
this isnt a language

jayvdb commented on bear-language-requirements.yaml:
more ... not a language .. remove

jayvdb commented on .ci/generate_bear_requirements.py:
```diff
@@ -168,6 +171,18 @@ def get_pip_requirements(requirements):
+            if(i=='All'):
+                value[n]=True
+        if 'default' in value:
+            value.remove('default')
```
@ayan-b , this is about `DocumentationStyleBear` which uses `DocstyleDefinition.get_available_definitions` - i think you were looking at this .. ?  It was caused by something in coalib ...?

Did you create an issue about it?

c.f. https://github.com/coala/coala-bears/pull/2688#discussion_r225751919

jayvdb commented on .ci/generate_bear_requirements.py:
```diff
@@ -168,6 +171,18 @@ def get_pip_requirements(requirements):
+    for value in language_dict.values():
+        for n,i in enumerate(value):
+            if(i=='All'):
+                value[n]=True
```
this is why `true` is going into the yaml .

At https://github.com/coala/coala-bears/pull/2688#discussion_r225752332 , I said that it should be `Bear: true` or `Bear: false`.  It could also be `Bear:`

It can not be
```yaml
Bear:
  - true
```

jayvdb commented on bear-requirements.yaml:
```diff
@@ -231,3 +231,4 @@ r_script_requirements:
     version: '>1.5'
   lintr:
     version: '>=1.0.2'
+    
```
why is this in your diff ????

li-boxuan replied:
@sladyn98 This is not a newcomer task, so you should be able to use `git ` properly, and people would not give you hints on `git` anymore. I remember in this PR, you add irrelevant stuff multiple times, and every time you say you don't add them 😅 Before you commit or push, verify your local diff is correct.

jayvdb replied:
This is still here.

before you push, run `git show` locally.  Read it, top to bottom.  Everything in there is your change.  If it doesnt belong there, do not push.  Fix it.

sladyn98 commented:
@jayvdb I did not make those changes

sladyn98 commented:
@jayvdb how to revert them?

jayvdb commented:
It wasnt magic :P  yes, you did it.  and I cant help today - im offline

sladyn98 commented:
Don't you guys think it's because I'm running the beat requirements script
that this file is getting changed?

On Sun, Oct 28, 2018, 9:43 AM John Vandenberg <notifications@github.com
wrote:

> It wasnt magic :P yes, you did it. and I cant help today - im offline
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coala/coala-bears/pull/2688#issuecomment-433674958>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbgGHsl4TqrkEfus0XGyjWOg2ORenPF7ks5upS7jgaJpZM4WbpAt>
> .
>

sladyn98 commented:
Yeah :( I will revert the changes as soon as possible. I am open to
learning as much as I can thanks.

jayvdb commented on .ci/generate_bear_requirements.py:
```diff
@@ -272,6 +297,14 @@ def sort_requirements(req_dict):
     else:
         output = open(args.output, 'w')
 
+    all_bears.sort(key = operator.attrgetter('name'))
```
this doesnt help, and also had whitespace around the `=`

jayvdb commented:
@gitmate-bot rebase

sladyn98 commented:
@jayvdb Okay cool . I learnt a lot from this since this was my first real code contribution and i made a lot of mistakes, but thanks to all the people who helped me @jayvdb @li-boxuan @ishanSrt @Naveenaidu for their input.Looking forward towards learning and contributing more to coala.

jayvdb commented:
@gitmate-bot ff
